impl_core/src/main/java/io/opencensus/implcore/trace/propagation/TraceContextFormat.java

```diff
@@ -0,0 +1,170 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Pattern;
+
+/*>>>
+import org.checkerframework.checker.nullness.qual.NonNull;
+*/
+
+/**
+ * Implementation of the TraceContext propagation protocol. See <a
+ * href=https://github.com/w3c/distributed-tracing>w3c/distributed-tracing</a>.
+ */
+public class TraceContextFormat extends TextFormat {
+  private static final Tracestate TRACESTATE_DEFAULT = Tracestate.builder().build();
+  @VisibleForTesting static final String TRACEPARENT = "traceparent";
+  @VisibleForTesting static final String TRACESTATE = "tracestate";
+  private static final List<String> FIELDS =
+      Collections.unmodifiableList(Arrays.asList(TRACEPARENT, TRACESTATE));
+
+  private static final String VERSION = "00";
+  private static final int VERSION_SIZE = 2;
+  private static final char TRACEPARENT_DELIMITER = '-';
+  private static final int TRACEPARENT_DELIMITER_SIZE = 1;
+  private static final int TRACE_ID_HEX_SIZE = 2 * TraceId.SIZE;
+  private static final int SPAN_ID_HEX_SIZE = 2 * SpanId.SIZE;
+  private static final int TRACE_OPTION_HEX_SIZE = 2 * TraceOptions.SIZE;
+  private static final int TRACE_ID_OFFSET = VERSION_SIZE + TRACEPARENT_DELIMITER_SIZE;
+  private static final int SPAN_ID_OFFSET =
+      TRACE_ID_OFFSET + TRACE_ID_HEX_SIZE + TRACEPARENT_DELIMITER_SIZE;
+  private static final int TRACE_OPTION_OFFSET =
+      SPAN_ID_OFFSET + SPAN_ID_HEX_SIZE + TRACEPARENT_DELIMITER_SIZE;
+  private static final int TRACEPARENT_HEADER_SIZE = TRACE_OPTION_OFFSET + 2 * TraceOptions.SIZE;
+  private static final String TRACEPARENT_HEADER_PREFIX = VERSION + TRACEPARENT_DELIMITER;
+  private static final int TRACESTATE_MAX_SIZE = 512;
+  private static final int TRACESTATE_MAX_MEMBERS = 32;
+  private static final char TRACESTATE_KEY_VALUE_DELIMITER = '=';
+  private static final char TRACESTATE_ENTRY_DELIMITER = ',';
+  private static final Splitter TRACESTATE_ENTRY_DELIMITER_SPLITTER =
+      Splitter.on(Pattern.compile("[ \t]*" + TRACESTATE_ENTRY_DELIMITER + "[ \t]*"));
+
+  @Override
+  public List<String> fields() {
+    return FIELDS;
+  }
+
+  @Override
+  public <C /*>>> extends @NonNull Object*/> void inject(
+      SpanContext spanContext, C carrier, Setter<C> setter) {
+    checkNotNull(spanContext, "spanContext");
+    checkNotNull(setter, "setter");
+    checkNotNull(carrier, "carrier");
+    setter.put(
+        carrier,
+        TRACEPARENT,
+        TRACEPARENT_HEADER_PREFIX
+            + spanContext.getTraceId().toLowerBase16()
+            + TRACEPARENT_DELIMITER
+            + spanContext.getSpanId().toLowerBase16()
+            + TRACEPARENT_DELIMITER
+            + spanContext.getTraceOptions().toLowerBase16());
+    List<Tracestate.Entry> entries = spanContext.getTracestate().getEntries();
+    if (entries.isEmpty()) {
+      // No need to add an empty "tracestate" header.
+      return;
+    }
+    StringBuilder stringBuilder = new StringBuilder(TRACESTATE_MAX_SIZE);
```

Comment:
does it mean that the last entry may be cut in the middle if total size overflows the limit?

Reply:
No, it pre-allocates a buffer of this size, if I go over the class will do another allocation which will have overhead. This is an optimization to make only one allocation for the buffer.